llmfoundry/models/hf/hf_causal_lm.py

```diff
@@ -77,196 +76,209 @@ def __init__(self, om_model_config: Union[DictConfig,
+                    trust_remote_code=trust_remote_code,
+                )
+        else:
+            raise ValueError(
+                f'init_device="{init_device}" must be either "cpu" or "meta".')
+
+        signal_file_path = f'.node_{dist.get_node_rank()}_local_rank0_completed'
+        if dist.get_local_rank() == 0:
+            with open(signal_file_path, 'wb') as f:
+                f.write(b'local_rank0_completed_download')
 
-            signal_file_path = f'.node_{dist.get_node_rank()}_local_rank0_completed'
-            if dist.get_local_rank() == 0:
-                with open(signal_file_path, 'wb') as f:
-                    f.write(b'local_rank0_completed_download')
+        # Avoid the collective call until the local rank zero has finished trying to download the checkpoint
+        # so that we don't timeout for large downloads. This syncs all processes on the node
+        with dist.local_rank_zero_download_and_wait(signal_file_path):
+            # Then, wait to ensure every node has finished downloading the checkpoint
+            dist.barrier()
 
-            # Avoid the collective call until the local rank zero has finished trying to download the checkpoint
-            # so that we don't timeout for large downloads. This syncs all processes on the node
-            with dist.local_rank_zero_download_and_wait(signal_file_path):
-                # Then, wait to ensure every node has finished downloading the checkpoint
-                dist.barrier()
+        if dist.get_local_rank() == 0:
+            os.remove(signal_file_path)
 
-            if dist.get_local_rank() == 0:
-                os.remove(signal_file_path)
+        z_loss = om_model_config.get('z_loss', 0.0)
 
-            z_loss = om_model_config.get('z_loss', 0.0)
+        attention_patch_type = om_model_config.get('attention_patch_type', None)
+        if attention_patch_type is not None:
+            if model.config.model_type != 'llama':
+                raise ValueError(
+                    f'attention_patch_type is only supported for llama models, but got {model.config.model_type}'
+                )
 
-            attention_patch_type = om_model_config.get('attention_patch_type',
-                                                       None)
-            if attention_patch_type is not None:
-                if model.config.model_type != 'llama':
+            log.debug(
+                f'Patching llama attention with {attention_patch_type} attention'
+            )
+            from transformers.models.llama.modeling_llama import LlamaAttention
+
+            from llmfoundry.models.layers.llama_attention_monkeypatch import \
+                get_llama_attention_patch_fn
+            LlamaAttention.forward = get_llama_attention_patch_fn(
+                attention_patch_type)
+            model.config.use_cache = False
+
+        if model.config.tie_word_embeddings and resolved_init_device == 'meta':
+            model.tie_weights()
```

Comment:
what does this do?

Reply:
will add comment. HF's weight tying doesn't apply properly when initing on meta device, so we have to manually tie the weights after init.